### PR TITLE
Docs/dependency docs

### PR DESCRIPTION
**Pull request by ardasener**
This PR adds documentation to some of the changes made in #153. Specifically it adds the following pages:

- [Getting Started / Optional Dependencies](https://github.com/sparcityeu/sparsebase/blob/docs/dependency_docs/docs/pages/getting_started/optional_dependencies.md)
- [Contributing / Adding New Dependencies](https://github.com/sparcityeu/sparsebase/blob/docs/dependency_docs/docs/pages/contributing/dependencies.md)
- [Contributing / Instantiating Classes](https://github.com/sparcityeu/sparsebase/blob/docs/dependency_docs/docs/pages/contributing/instantiations.md)

### CODE REVIEW

File: docs/pages/getting_started/optional_dependencies.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+Optional dependencies in the context of sparsebase are libraries that for one reason 
+or another can not be directly bundled sparsebase. 
+
+## Optional Dependency List
+
+- [METIS](https://github.com/KarypisLab/METIS)
+
+## Compiling With An Optional Dependency
+
+To enable an optional dependency the user must either pass `-DUSE_<dependency>=ON` parameter to cmake.
```
Comment (AmroAlJundi) [inline docs/pages/getting_started/optional_dependencies.md:12]: to cmake, or set ...

File: docs/pages/contributing/dependencies.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+- [PATOH](https://faculty.cc.gatech.edu/~umit/software.html) (Not open-source)
+
+> If the dependency is header-only we suggest treating it as a required dependency 
+> even if it doesn't have substantial effect on the library
+
+Optional dependencies must be build separately by users.
+
+Here are the steps to add an optional dependency:
+
+1. In the `src/CMakeLists.txt` file add a new option in the form `USE_<dependency>`.
```
Comment (AmroAlJundi) [inline docs/pages/contributing/dependencies.md:52]: All of these items show up as (1) when you compile the page. Maybe there's too much space between them?

File: docs/pages/contributing/dependencies.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Adding New Dependencies
+
+SparseBase contains two different types of dependencies:
+- Required dependencies
```
Comment (AmroAlJundi) [inline docs/pages/contributing/dependencies.md:4]: Very nitpicky but can you add periods at the end of bullet point sentences :)

File: docs/pages/contributing/instantiations.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+As a result of this, we try to keep the header files as light as possible by avoiding putting implementations there.
+There are a few exceptions to this (like the `As<>()` function of `Format`) 
+but this is only done in very rare situations.
+
+Hence, any templated class or function you create in SparseBase will require instantiation as it needs to be compiled
+with the library with the types known at compile time. This is achieved with the script 
+`src/generate_explicit_instantiations.py`.
+
+> Note that this script is not intended to be called manually besides debugging. It will be called automatically
```
Comment (AmroAlJundi) [inline docs/pages/contributing/instantiations.md:14]: For some reason this quote didn't show up on my computer. Just looks indented.
resim

File: docs/pages/contributing/instantiations.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+2. The identifiers starting with `$` will be replaced with their counterparts in the CMake variables. Currently, we have
+`$id_type`, `$nnz_type`, `$value_type`, `$float_type`. Every possible combination will be covered.
+
+
+## Filename
+
+The filename to be used for the output. As per [Google's Style Guide](https://google.github.io/styleguide/cppguide.html),
+the filenames we use end with the `.inc` extension. 
+
+> Caution!!! If the file is being used for the first time you should also call the `reset_file()` function with it.
```
Comment (AmroAlJundi) [inline docs/pages/contributing/instantiations.md:39]: Same with above. Could be my doxygen idk

**Review by AmroAlJundi — CHANGES_REQUESTED** [review]
The numbered list in one page is not working and quotes don't show up the way you'd expect. Plus some missing periods. Otherwise very nice. Only the first change is important btw everything else super optional.

File: docs/pages/contributing/instantiations.md (hunk 4 above)
Comment (ardasener) [reply docs/pages/contributing/instantiations.md:14]: Same on mine, I don't think we can fix that :(

File: docs/pages/contributing/dependencies.md (hunk 2 above)
Comment (ardasener) [reply docs/pages/contributing/dependencies.md:52]: Having code blocks in between breaks it. Indenting them should fix it apparently, I will try that.

File: docs/pages/contributing/instantiations.md (hunk 5 above)
Comment (ardasener) [reply docs/pages/contributing/instantiations.md:39]: Not related with doxygen. Read The Docs theme just defines them like that. Sphinx has dedicated warning and info blocks I will use them instead.

Comment (ardasener) [reply docs/pages/contributing/instantiations.md:39]: Also this is not local to this page. For example our current live [getting started page](https://sparcityeu.github.io/sparsebase/pages/getting_started.html) has the same issue.

**Review by AmroAlJundi — APPROVED** [review]

**Pull request merged**